keps/sig-api-machinery/1040-priority-and-fairness/README.md

```diff
@@ -1066,8 +1076,361 @@ the remaining requests in that queue start getting faster service.  In
+to just WATCH requests.
+
+However, even for WATCH requests, there are effectively two separate problems
+that has to be considered and addressed.
+
+#### Watch initialization
+
+While this is an important piece to address, to allow incremental progress
+we are leaving addressing this problem for the future release.
+
+<<[UNRESOLVED]>>
+The most compatible approach would be to simply include WATCH requests as
+ones that are processed by APF dispatcher, but allow them to send an artificial
+`request finished` signal to the dispatcher once their initialization is done.
+However, that requires more detailed design.
+<<[/UNRESOLVED]>>
+
+#### Keeping the watch up-to-date
+
+Once the watch is initialized, we have to keep it up-to-date with incoming
+changes.  That basically means that whenever some object that this particular
+watch is interested in is added/updated/deleted, and appropriate event has
+to be sent.
+
+Similarly as in generic case of long-running requests, predicting this cost
+up front is impossible. Fortunately, in this case we have a full control over
+those events, because those are effectively a result of the mutating requests
+that out mechanism is explicitly admitting.
+So instead of associating the cost of sending watch events to the corresponding
+WATCH request, we will reverse the situation and associate the cost of sending
+watch events to the mutating request that triggering them.
+
+In other words, we will be throttling the write requests to ensure that
+apiserver would be able to keep up with the watch traffic instead of throttling
+the watchers themselves.  The main reason for that is that throttling watchers
+themselves isn't really effective: we either need to send all objects to them
+anyway or in case we close them, they will try to resume from last received
+event anyway.  Which means we don't get anything on throttling them.
+
+##### Estimating cost of the request
+
+Let's start with an assumption that sending every watch event is equally
+expensive.  We will discuss how to generalize it below.
+
+With the above assumption, a cost of a mutating request associated with
+sending watch events triggerred by it is proportional to the number of
+watchers that has to process that event.   So let's describe how we can
+estimate this number.
+
+Obviously, we can't afford going over all watches to compute that - we need
+to keep this information already precomputed. What if we would simply
+store an in-memory map from (resource type, namespace, name) tuple into
+number of opened watches that are interested in a given event.  The size of
+that map won't be larger then the total number of watches, so that is
+acceptable.
+Note that each watch can also specify label and field selectors.  However,
+in most of cases a particular object has to be processed for them anyway
+to check if the selectors are satisfied.  So we can ignore those selectors
+as the object contributes to the cost (even if it will not be send as not
```

Comment:
There are different costs.  But to a first approximation this may be good enough.